File: fig5.py
import numpy as np
import matplotlib.pyplot as plt
from IsingRG import exact_UC

K,U1,C1 = np.loadtxt('fig5_L32.txt').T
K,U2,C2 = np.loadtxt('fig5_L128.txt').T
Ke = np.linspace(K[0], K[-1], 256)
Ue,Ce = exact_UC(Ke)

plt.figure(figsize=(6.4,6))
plt.subplot(211)
plt.plot(K,U1,'.-', label=r'$L=32$')
plt.plot(K,U2,'.-', label=r'$L=128$')
plt.plot(Ke,Ue, label='exact')
plt.ylabel(r'$u$ = internal energy', fontsize=14)
plt.legend(fontsize=14)

plt.subplot(212)
plt.plot(K,C1,'.-', label=r'$L=32$')
plt.plot(K,C2,'.-', label=r'$L=128$')
plt.plot(Ke,Ce, label='exact')
plt.ylabel(r'$c$ = specific heat', fontsize=14)
plt.xlabel(r'$K = J/(k_{\rm B}T)$', fontsize=14)
plt.legend(fontsize=14)

plt.tight_layout()
plt.savefig('fig5.eps')
plt.show()

File: fig5.cpp
#include "IsingMC.h"
#include<iostream>
#include<fstream>

void fig5(const char *fname, long L,
          double K1, double K2, long n) {
    std::ofstream f(fname);
    double dK((K2-K1)/n),K,U,C;
    long k;
    IsingMC m(L);
    for(k=0; k<=n; k++) {
        K = K1 + k*dK;
        m.set_K(K,1<<10);
        m.get_UC(U,C,1<<16);
        f << K << ' ' << U << ' ' << C << std::endl;
        std::cout << K << ' ' << U << ' ' << C << std::endl;
    }
}

main() {
    fig5("fig5_L32.txt", 32, 0.4, 0.5, 40);
    fig5("fig5_L128.txt", 128, 0.4, 0.5, 40);
}